packages/cdktf/lib/terraform-asset.ts

```diff
@@ -0,0 +1,112 @@
+import { Construct, ISynthesisSession, Node } from "constructs";
+import * as fs from "fs";
+import * as path from "path";
+import { Manifest } from "./manifest";
+import { copySync, archiveSync, hashPath } from "./private/fs";
+import { Resource } from "./resource";
+
+export interface TerraformAssetConfig {
+  readonly path: string;
+  readonly type?: AssetType;
+  readonly assetHash?: string;
+}
+
+export enum AssetType {
+  FILE,
+  DIRECTORY,
+  ARCHIVE,
+}
+
+const ARCHIVE_NAME = "archive.zip";
+const ASSETS_DIRECTORY = "assets";
+
+export class TerraformAsset extends Resource {
+  private sourcePath: string;
+  public assetHash: string;
+  public type: AssetType;
+
+  constructor(scope: Construct, id: string, config: TerraformAssetConfig) {
+    super(scope, id);
+
+    if (!path.isAbsolute(config.path)) {
+      throw new Error(
+        `TerraformAsset path needs to be absolute, got relative path: '${config.path}'`
+      );
+    }
+
+    const stat = fs.statSync(config.path);
+    const inferredType = stat.isFile() ? AssetType.FILE : AssetType.DIRECTORY;
+    this.type = config.type ?? inferredType;
+    this.sourcePath = config.path;
+    this.assetHash = config.assetHash || hashPath(this.sourcePath);
+
+    if (stat.isFile() && this.type !== AssetType.FILE) {
+      throw new Error(
+        `TerraformAsset ${id} expects path to be a directory, a file was passed: '${config.path}'`
+      );
+    }
+
+    if (!stat.isFile() && this.type === AssetType.FILE) {
+      throw new Error(
+        `TerraformAsset ${id} expects path to be a file, a directory was passed: '${config.path}'`
+      );
+    }
+  }
+
+  public get path(): string {
+    return path.join(
+      ASSETS_DIRECTORY,
+      this.stack.getLogicalId(Node.of(this)), // needed to get a human friendly, unique segment in the path
+      this.fileName ?? ""
+    );
+  }
+
+  public get fileName(): string | undefined {
+    switch (this.type) {
+      case AssetType.ARCHIVE:
+        return ARCHIVE_NAME;
+      case AssetType.FILE:
+        return path.basename(this.sourcePath);
+      case AssetType.DIRECTORY:
+        return undefined;
```

Comment:
Perhaps fail loud and explicitly in this case?

Reply:
I wasn't considering this a failure, undefined signals for me that a value does not exist. Do you think people might run into problems here?

Reply:
If a user tries to get a filename for the directory, what would you think is the intention? I'd assume the user is trying to get some recognizable name. Now, I think you're right - this shouldn't result in an error. Perhaps we're more looking for something like basename?

```ts
> path.basename('/tmp/foo/bar')
'bar'
> path.basename('/tmp/foo/bar.zip')
'bar.zip'
```

Reply:
I really did not have a use-case in mind, but yours makes sense, basename fits fine here 👍